core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java

```diff
@@ -111,49 +162,228 @@ public FieldSortBuilder sortMode(String sortMode) {
         }
         if (nestedFilter != null) {
-            builder.field("nested_filter", nestedFilter, params);
+            builder.field(NESTED_FILTER.getPreferredName(), nestedFilter, params);
         }
         if (nestedPath != null) {
-            builder.field("nested_path", nestedPath);
+            builder.field(NESTED_PATH.getPreferredName(), nestedPath);
         }
         builder.endObject();
         return builder;
     }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true;
+        }
+
+        if (other == null || getClass() != other.getClass()) {
+            return false;
+        }
+
+        FieldSortBuilder builder = (FieldSortBuilder) other;
+        return (Objects.equals(this.fieldName, builder.fieldName) && Objects.equals(this.nestedFilter, builder.nestedFilter)
+                && Objects.equals(this.nestedPath, builder.nestedPath) && Objects.equals(this.missing, builder.missing)
+                && Objects.equals(this.order, builder.order) && Objects.equals(this.sortMode, builder.sortMode)
+                && Objects.equals(this.unmappedType, builder.unmappedType));
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.fieldName, this.nestedFilter, this.nestedPath, this.missing, this.order, this.sortMode, this.unmappedType);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(this.fieldName);
+        if (this.nestedFilter != null) {
+            out.writeBoolean(true);
+            out.writeQuery(this.nestedFilter);
+        } else {
+            out.writeBoolean(false);
+        }
+        out.writeOptionalString(this.nestedPath);
+        out.writeGenericValue(this.missing);
+
+        if (this.order != null) {
+            out.writeBoolean(true);
+            this.order.writeTo(out);
+        } else {
+            out.writeBoolean(false);
+        }
+
+        out.writeOptionalString(this.sortMode);
+        out.writeOptionalString(this.unmappedType);
+    }
+
+    @Override
+    public FieldSortBuilder readFrom(StreamInput in) throws IOException {
+        String fieldName = in.readString();
+        FieldSortBuilder result = new FieldSortBuilder(fieldName);
+        if (in.readBoolean()) {
+            QueryBuilder query = in.readQuery();
+            result.setNestedFilter(query);
+        }
+        result.setNestedPath(in.readOptionalString());
+        result.missing(in.readGenericValue());
+
+        if (in.readBoolean()) {
+            result.order(SortOrder.readOrderFrom(in));
+        }
+        result.sortMode(in.readOptionalString());
+        result.unmappedType(in.readOptionalString());
+        return result;
+    }
+
+    @Override
+    public FieldSortBuilder fromXContent(QueryParseContext context, String elementName) throws IOException {
```

Comment:
Yes, sorry for the confusion, I remember the discussion now. Maybe just rename then, although `elementName` is also fine.